events/events_test.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package events
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+)
+
+type testEvent struct{}
+
+type EventsTestSuite struct {
+	suite.Suite
+
+	e EventEmitter
+}
+
+func (s *EventsTestSuite) TestEventListenerRegistration() {
+	var wg1 sync.WaitGroup
+	var wg2 sync.WaitGroup
+
+	// expect listener to be invoked only once
+	wg1.Add(1)
+	l1 := &MockEventListener{}
+	l1.On("HandleEvent", testEvent{}).Run(func(args mock.Arguments) {
+		wg1.Done()
+	}).Return()
+	s.e.AddListener(l1)
+	// make sure listener does stick around for next tests
+	defer s.e.RemoveListener(l1)
+
+	// expect listener to be invoked twice
+	wg2.Add(2)
+	l2 := &MockEventListener{}
+	l2.On("HandleEvent", testEvent{}).Run(func(args mock.Arguments) {
+		wg2.Done()
+	}).Return()
+	s.e.AddListener(l2)
+	// make sure listener does stick around for next tests
+	defer s.e.RemoveListener(l2)
+
+	// emit event by value
+	s.e.EmitEvent(testEvent{})
+
+	wg1.Wait()
```

Comment:
In my opinion a never ending test suite signals to the user well enough that a regression has happened and when ran in verbose mode you will see in which test a deadlock happens.

The problem with such timeouts is that it adds to flappy tests during race detection (and also without the race detector but they are found more often with the race detector). I have removed all timeout's during our tests to make it pass the race detector test consistently so I'm not keen on adding one here again.